core/src/main/scala/org/http4s/Uri.scala

```diff
@@ -279,6 +275,126 @@ object Uri {
     }
   }
 
+  final class Path private (
+      val segments: List[Path.Segment],
+      val absolute: Boolean,
+      val endsWithSlash: Boolean)
+      extends Renderable {
+
+    def isEmpty: Boolean = segments.isEmpty
+    def nonEmpty: Boolean = segments.nonEmpty
+
+    override def equals(obj: Any): Boolean = obj match {
+      case p: Path => doEquals(p)
+      case _ => false
+    }
+
+    private def doEquals(path: Path): Boolean =
+      this.segments == path.segments && path.absolute == this.absolute && path.endsWithSlash == this.endsWithSlash
+
+    override def hashCode(): Int = {
+      var hash = segments.hashCode()
+      hash += 31 * java.lang.Boolean.hashCode(absolute)
+      hash += 31 * java.lang.Boolean.hashCode(absolute)
+      hash
+    }
+
+    def render(writer: Writer): writer.type = {
+      val start = if (absolute) "/" else ""
+      writer << start << segments.iterator.mkString("/")
+      if (endsWithSlash) writer << "/" else writer
+    }
+
+    override val renderString: String = super.renderString
+    override def toString: String = renderString
+
+    def /(segment: Path.Segment): Path = addSegment(segment)
+    def addSegment(segment: Path.Segment): Path =
+      addSegments(List(segment))
+    def addSegments(value: List[Path.Segment]): Path =
+      Path(this.segments ++ value, absolute = absolute || this.segments.isEmpty)
+
+    def normalize: Path = Path(segments.filterNot(_.isEmpty))
+
+    /* Merge paths per RFC 3986 5.2.3 */
+    def merge(path: Path): Path = {
+      val merge = segments.init
+      Path(merge ++ path.segments, absolute = absolute, endsWithSlash = path.endsWithSlash)
+    }
+
+    def startsWith(path: Path): Boolean = segments.startsWith(path.segments)
+
+    def startsWithString(path: String): Boolean = startsWith(Path.fromString(path))
+
+    def indexOf(path: Path): Option[Int] =
+      Some(segments.indexOfSlice(path.segments)).filterNot(_ == -1)
+
+    def indexOfString(path: String): Option[Int] = indexOf(Path.fromString(path))
+
+    def splitAt(idx: Int): (Path, Path) = {
+      val (start, end) = segments.splitAt(idx)
+      Path(start, absolute = absolute) -> Path(end, endsWithSlash = endsWithSlash)
+    }
+  }
+
+  object Path {
+    val empty = Path(Nil)
+    val Root = Path(Nil, absolute = true)
+
+    class Segment private (val encoded: String) {
+      def isEmpty = encoded.isEmpty
+
+      override def equals(obj: Any): Boolean = obj match {
+        case s: Segment => s.encoded == encoded
+      }
+
+      override def hashCode(): Int = encoded.hashCode
+
+      override val toString: String = encoded
+    }
+
+    object Segment extends (String => Segment) {
```

Comment:
We miss this function on a lot of case-classlike-things-that-aren't-case-classes.